src/components/SuggestionsList/SuggestionsList.tsx

```diff
@@ -0,0 +1,158 @@
+/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
+import '../../yamui';
+import * as React from 'react';
+import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
+import SuggestionsListItem, { SuggestionItem } from './SuggestionsListItem';
+import Spinner from '../Spinner';
+import Text, { TextSize, TextColor } from '../Text';
+import Block, { GutterSize } from '../Block';
+import './SuggestionsList.css';
+
+export interface SuggestionsListProps extends NestableBaseComponentProps {
+  /**
+   * Top positioning of the list in pixels.
+   */
+  top: number;
+  /*
+   * Left positioning of the list in pixels.
+   */
+  left: number;
+  /**
+   * Left positioning of the list in pixels.
+   * @default 'left'
+   */
+  isLoading?: boolean;
+  /**
+   * Text to show next to loading spinner.
+   */
+  loadingText: string;
+  /**
+   * Text to show when the results are empty for the active search.
+   */
+  noResultsText: string;
+  /**
+   * The groups of items to render.
+   */
+  groupedItems?: SuggestionItemGroupProps[];
+  /**
+   * The index of current keyboard selection.
+   */
+  selectedIndex: number;
```

Comment:
selectedId would be easier for the client to deal with when there are multiple groups of results. but it'd still be nice if the client didn't have to deal with the selection at all. would it be possible for SuggestionsList to accept a target prop which is a ref to the input or textarea, and listen to its keyboard interaction directly?

Reply:
So in a Draft world there's various props - onUpArrow, onDownArrow, handleReturn etc - how would it work?

Reply:
is there a way to get the ref to Draft's underlying component? haven't verified but might be possible?
https://github.com/facebook/draft-js/issues/1398#issuecomment-331047429
https://jsfiddle.net/js1vn56m/

Reply:
if so, could pass that ref as SuggestionList's input target and listen to the events directly

Reply:
you don't want to be in the business of intercepting or interfacing with the DOM directly - that's the whole point in Draft. Moreover, Draft's event handling for return is handleReturn rather than onReturn - because it expects you to return a 'handled' vs 'not-handled' string

Reply:
Maybe we can create a DOM element on mount, move it on render, and attach the callout to that element?

Reply:
sketching some ideas up and will report back

Reply:
https://github.int.yammer.com/gist/rtagger/645580bfd6de1f36bc11e31ddc929139 thoughts on this?

Reply:
Draft says the up/down keys are non-cancelable:
https://github.com/facebook/draft-js/blob/master/src/component/base/DraftEditorProps.js#L156
If listening to those events on a Draft instance to handle up/down in SuggestionsList, does that mean the cursor position will be bouncing around in the textarea as well? if so, that's an argument for passing the DOM node ref to SuggestionsList to listen and cancel events while SuggestionsList is rendered.


Regarding selectedId vs. selectedIndex. Draft won't know or care about SuggestionsList or its selected Suggestion, your component is only using Draft APIs to place the SuggestionsList over the cursor and then inject the selected item into the editor. your component would be managing the data powering the list of suggestions and selected state. i'm suggesting to select by id instead of index because your data model powering the SuggestionsList could change at any point and it would be easier to make sure the selected item remains selected if you select it by id, especially once we implement grouped suggestions.

but i don't think we need selectedIndex/selectedItem at all if this isn't a controlled component. if we can pass SuggestionsList the input DOM element then it can listen to up/down keys itself and the parent component can just listen to the onItemSelected() callback prop

Reply:
even though it says non-cancelable it passes the event object, so maybe you still can preventDefault()

Reply:
Yeah there was no jumping around, I prevent defaulted here https://github.int.yammer.com/yammer/yamjs/pull/9494/files#diff-551493f5e3d5008270b2ae82f390f960R97